mlflow/diviner.py

```diff
@@ -0,0 +1,452 @@
+        predict_groups = conf.get("groups", None)
+
+        if predict_groups and not isinstance(predict_groups, List):
+            raise MlflowException(
+                "Specifying a group subset for prediction requires groups to be defined as a "
+                f"[List[Tuple|List[<group_keys>]]. Submitted group type: {type(predict_groups)}."
+            )
+
+        # NB: json serialization of a tuple converts the tuple to a List. Diviner requires a
+        # List of Tuples to be input to the group_prediction API. This conversion is for utilizing
+        # the pyfunc flavor through the serving API.
+        if predict_groups and not isinstance(predict_groups[0], Tuple):
+            predict_groups = [tuple(group) for group in predict_groups]
+
+        if self.diviner_instance_type == "GroupedProphet":
+
+            # We're wrapping two different endpoints to Diviner here for the pyfunc implementation.
+            # Since we're limited by a single endpoint, we can address redirecting to the
+            # method ``predict_groups()`` which will allow for a subset of groups to be forecasted
+            # if the prediction configuration DataFrame contains a List[tuple[str]]] in the
+            # ``groups`` column. If this column is not present, all groups will be used to generate
+            # forecasts, utilizing the less computationally complex method ``forecast``.
+            if not predict_groups:
+                prediction_df = self.diviner_model.forecast(horizon=n_periods, frequency=frequency)
+            else:
+                group_kwargs = {k: v for k, v in conf.items() if k in {"predict_col", "on_error"}}
+                prediction_df = self.diviner_model.predict_groups(
+                    groups=predict_groups, horizon=n_periods, frequency=frequency, **group_kwargs
+                )
+
+            if predict_col is not None:
+                prediction_df.rename(columns={"yhat": predict_col}, inplace=True)
+
+        elif self.diviner_instance_type == "GroupedPmdarima":
+
+            # As above, we're redirecting the prediction request to one of two different methods
+            # for ``Diviner``'s pmdarima implementation. If the ``groups`` column is present with
+            # a list of tuples of keys to lookup, ``predict_groups()`` will be used. Otherwise,
+            # the standard ``predict()`` method will be called to generate forecasts for all groups
+            # that were trained on.
+            restricted_keys = {"n_periods", "horizon", "frequency", "groups"}
+
+            predict_conf = {k: v for k, v in conf.items() if k not in restricted_keys}
+            if not predict_groups:
+                prediction_df = self.diviner_model.predict(n_periods=n_periods, **predict_conf)
+            else:
+                prediction_df = self.diviner_model.predict_groups(
+                    groups=predict_groups, n_periods=n_periods, **predict_conf
+                )
+        else:
```

Comment:
All bets are off on other libraries that may get wrapped in the future. I figured that version-locking support for implementations is the safer route to reduce user confusion.